### PR TITLE
CI: Support for Python 3.9

### PR DESCRIPTION
**Pull request by csadorf**
According to the [Python 3.9 RC release notes](https://www.python.org/downloads/release/python-390rc1/):

> We strongly encourage maintainers of third-party Python projects to prepare their projects for 3.9 compatibility during this phase. As always, report any issues to the Python bug tracker.

I will track preparations to support Python 3.9 on this branch. The CI workflow for this branch is expected to fail until further notice.

### CODE REVIEW

**Comment by sphuber** [comment]
Just to note that [Python 3.9 was officially released](https://www.python.org/downloads/release/python-390/) today. Should probably wait for our dependencies to release compatible versions, but hopefully we can also soon add support

**Comment by csadorf** [comment]
> Just to note that [Python 3.9 was officially released](https://www.python.org/downloads/release/python-390/) today. Should probably wait for our dependencies to release compatible versions, but hopefully we can also soon add support

I'll update this PR.

**Comment by csadorf** [comment]
Blocked by https://github.com/aiidateam/aiida-core/pull/4422 .

**Comment by csadorf** [comment]
Blocked by https://github.com/actions/setup-python/issues/148 .

**Comment by csadorf** [comment]
Standard pip install succeeds now, installation with extras still fails. Python 3.9 is not available via conda channels yet, but should be shortly.

**Comment by csadorf** [comment]
Conda can now be installed via the conda-forge channel, but there are still many conflicts, likely because upstream packages are not available yet for Python 3.9.

**Comment by csadorf** [comment]
The pip install with extras is currently blocked by: https://github.com/nucleic/kiwi/issues/92

**Comment by sphuber** [comment]
Just realized that all of this is also completely dependent on our own maintained dependencies releasing a Python 3.9 compatible version: `reentry`, `kiwipy`, `plumpy` and maybe some others I am now forgetting. I will start working on the latter two.

**Comment by csadorf** [comment]
The conda install now passes, pip install with extensions still failing, because of kiwisolver: https://github.com/nucleic/kiwi/issues/92

A release of a Python 3.9 compatible version is promised for by the end of the week.

**Comment by csadorf** [comment]
The pip install with and without extras succeeds now, but the conda installation no longer passes (go figure)..

**Comment by sphuber** [comment]
> The pip install with and without extras succeeds now, but the conda installation no longer passes (go figure)..

I am bit surprised that it passes at all to be honest. We haven't released a Python 3.9 compatible version for `plumpy` nor `kiwipy` yet. Granted, it may just be that the 3.8 version happens to be 3.9 compatible, but shouldn't pip automatically check if the package is explicitly marked to be compatible with 3.9? Or apparently it will always just try to install it?

**Comment by chrisjsewell** [comment]
> I am bit surprised that it passes at all to be honest. We haven't released a Python 3.9 compatible version for `plumpy` nor `kiwipy` yet. Granted, it may just be that the 3.8 version happens to be 3.9 compatible, but shouldn't pip automatically check if the package is explicitly marked to be compatible with 3.9? Or apparently it will always just try to install it?

Well thats because of your loose python pinning 😜 https://github.com/aiidateam/kiwipy/blob/7f6f642dcc527733ea57604e72e192afa6c6fd97/setup.py#L32

**Comment by sphuber** [comment]
> > I am bit surprised that it passes at all to be honest. We haven't released a Python 3.9 compatible version for `plumpy` nor `kiwipy` yet. Granted, it may just be that the 3.8 version happens to be 3.9 compatible, but shouldn't pip automatically check if the package is explicitly marked to be compatible with 3.9? Or apparently it will always just try to install it?
> 
> Well thats because of your loose python pinnings stuck_out_tongue_winking_eye
> https://github.com/aiidateam/plumpy/blob/87337f649fe2b03e7007301b36e2633b6918fc4b/setup.py#L30
> https://github.com/aiidateam/kiwipy/blob/7f6f642dcc527733ea57604e72e192afa6c6fd97/setup.py#L32

A well, fair enough. I forgot about that, I was thinking whether `pip` also looked at the `'Programming Language :: Python :: 3.8'` classifiers, but I guess it doesn't limit based on that.

**Comment by chrisjsewell** [comment]
> A well, fair enough. I forgot about that, I was thinking whether pip also looked at the 'Programming Language :: Python :: 3.8' classifiers, but I guess it doesn't limit based on that.

No, I've been guilty of it as well, but you should really pin `python_requires` to the same as the classifiers denote

**Comment by chrisjsewell** [comment]
same with this repo actually: https://github.com/aiidateam/aiida-core/blob/9ff07c166a559f98b5b2be71537814ec00d3f18d/setup.json#L10

**Comment by sphuber** [comment]
> > A well, fair enough. I forgot about that, I was thinking whether pip also looked at the 'Programming Language :: Python :: 3.8' classifiers, but I guess it doesn't limit based on that.
> 
> No, I've been guilty of it as well, but you should really pin `python_requires` to the same as the classifiers denote

Isn't there a tradeoff? Clearly by adding an upper limit, you don't run the risk that the package will break for a new Python release. The same would happen with an upper limit, there it would just not even install and so the end result is the same, that package cannot be used in the new version. However, the advantage of not adding the upper limit is that there is a chance (and it being minor versions, they should be forwards compatible and therefore guaranteed) that the existing package simply works with the new version. This means upstream does not have to wait for a release. Imagine if every library put an upper limit. Each environment would then have to wait for everything downstream to have released a new version before it could do anything, massively slowing down migration to new Python releases.

I have never thought or read about this so this is an honest question on what the "best" practice is here.

**Comment by chrisjsewell** [comment]
> I have never thought or read about this so this is an honest question on what the "best" practice is here.

Yeh neither have I really.
I guess TBF all v3.x should be back compatible, so there should never be anything that breaks when you test against a new major version

**Comment by csadorf** [comment]
I think it depends a bit on the complexity of the library/application. In the case of aiida-core it might make sense to set an upper limit and only increase it once it has been properly tested for that newer version. However, I would probably advocate for the "opposite" approach and only explicitly limit if incompatibilities are found.

**Review by sphuber — COMMENTED** [review]
Thanks @csadorf . Mostly have a few questions, some of which are in the code. Another one is that I am a bit surprised the tests actually pass. I have been seeing deprecation warnings coming from `reentry` for a while which still uses `collections.Mapping` which is deprecated and no longer works ion 3.9 in favor of `collections.abc.Mapping`. I guess that part of the code is not actually used in our code paths. @ltalirz I also found that `reentry` is still on dropd's fork. Should we move this to `aiidateam` and start to add other core members as maintainers in addition to yourself?

File: .github/workflows/test-install.yml
Comment (csadorf) [reply .github/workflows/test-install.yml:31]: No, the start-up time is quick here, it's only when you need to install aiida-core with extras that takes significantly longer.

File: .github/workflows/test-install.yml
Comment (csadorf) [reply .github/workflows/test-install.yml:95]: Updating conda fails for unknown reasons and I see no specific reason to keep it enabled.

**Comment by csadorf** [comment]
@sphuber I addressed your comments.

File: .github/workflows/test-install.yml
Comment (sphuber) [reply .github/workflows/test-install.yml:95]: Maybe make a note of this in the commit message when you squash merge please.

**Review by chrisjsewell — APPROVED** [review]
looks good thanks

File: .github/workflows/test-install.yml
Comment (sphuber) [reply .github/workflows/test-install.yml:95]: Actually, I will approve and merge it, don't worry about it

**Review by sphuber — APPROVED** [review]
Looking good, thanks @csadorf

**Comment by csadorf** [comment]
> Looking good, thanks @csadorf

Thx! I'll wait for all tests to pass and then squash-merge.

**Pull request merged**
